### ISSUE TITLE
Custom VJP with jax.custom_vjp returns incorrect gradient for a chained function

### ISSUE THREAD

**Issue by qdevpsi3** [google/jax#17496]
### Description:

When implementing a custom VJP for a particular operation (using matrix-vector multiplication as an illustrative example), I have encountered a peculiar gradient behavior in chained operations. Specifically, while the gradient is correctly computed for one input when the custom operation is applied in isolation, it fails to compute the gradient accurately in the context of chained operations.

Using the matrix-vector multiplication as an example: 
- The custom VJP was designed to compute the gradient only concerning matrices, leaving out vectors. 
- For a standalone application of this custom operation, gradients are evaluated correctly. 
- However, when this operation is part of a sequence (for instance, two matrix multiplications separated by non-linearities), the gradient for the first matrix is observed to be zero, while the gradient for the second matrix is computed correctly. 
- It's noteworthy that this occurs even though there was no need for a gradient with respect to the input vector in the chained sequence.

This behavior suggests potential issues with how gradients are propagated through chained operations when a custom VJP is involved.

Below is a minimal example to reproduce the issue:

```python
import jax
import jax.numpy as jnp

@jax.custom_vjp
def fn(matrix, vector):
    return jnp.dot(matrix, vector)
    
def fn_fwd(matrix, vector):
    return fn(matrix, vector), (matrix, vector)

def fn_bwd(res, g):
    matrix, vector = res
    matrix_dot = jnp.outer(g, vector)
    # The following gradient should not be needed for the example
    vector_dot = None # jnp.dot(matrix.T, g) 
    return matrix_dot, vector_dot

fn.defvjp(fn_fwd, fn_bwd)

custom_dot = fn

x = jnp.ones((2,))
A = B = jnp.ones((2, 2))
params = (A, B)

@jax.grad
def func(params, x):
    x = jax.nn.relu(jnp.dot(params[0], x))
    x = jax.nn.relu(jnp.dot(params[1], x))
    return x.mean()

print('Expected Behavior')
grads = func(params, x)
print(grads)  # This provides the expected output

@jax.grad
def func_with_custom_dot(params, x):
    x = jax.nn.relu(custom_dot(params[0], x))
    x = jax.nn.relu(custom_dot(params[1], x))
    return x.mean()

grads = func_with_custom_dot(params, x)
print('Actual Behavior')
print(grads)  # This does not provide the expected output
```

### Expected Behavior:

Both matrices A and B should have their gradients computed correctly.
```
(Array([[1., 1.],
       [1., 1.]], dtype=float64), Array([[1., 1.],
       [1., 1.]], dtype=float64))
```
### Actual Behavior:

The gradient for matrix A is zero while the gradient for matrix B is correct.
```
Actual Behavior
(Array([[0., 0.],
       [0., 0.]], dtype=float64), Array([[1., 1.],
       [1., 1.]], dtype=float64))
```

### What jax/jaxlib version are you using?

jax v0.4.14

### Which accelerator(s) are you using?

CPU

### Additional system info

Python 3.10, MacOS with M2 chip

**Comment by mattjj** [comment]
Thanks for the question!

This isn't a JAX bug (indeed the propagation rules for custom_vjps are the same as for everything else, and autodiff is perhaps the best-tested part of JAX), but just a bug in the `custom_vjp` rule. The issue is that this isn't accurate:
```python
    # The following gradient should not be needed for the example
    vector_dot = None # jnp.dot(matrix.T, g) 
```

Indeed if you just put that commented-out `jnp.dot(matrix.T, g)` back in, you get the correct behavior, as you may have noticed.

If you don't propagate gradients back to the "data" (rather than "params") side, you've effectively disconnected the first `custom_dot` from the output. Indeed with that `vector_dot = None` you've effectively written this:

```python
@jax.grad
def func_with_custom_dot(params, x):
    x = jax.nn.relu(custom_dot(params[0], x))
    x = jax.lax.stop_gradient(x)  # NOTE disconnecting!
    x = jax.nn.relu(custom_dot(params[1], x))
    return x.mean()
```

So I think perhaps this is just a bug in understanding why your rule needs to propagate gradients on both sides.

**Comment by mattjj** [comment]
Indeed if you change the reference function to read:

```python
@jax.grad
def func(params, x):
    x = jax.nn.relu(jnp.dot(params[0], x))
    x = jax.lax.stop_gradient(x)  # NOTE added this
    x = jax.nn.relu(jnp.dot(params[1], x))
    return x.mean()
```

you'll get the same result as the `custom_dot` version using `vector_dot = None`.
